osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using OpenTK.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Graphics.Sprites;
+using System;
+using System.Collections.Generic;
+
+namespace osu.Game.Graphics.Containers
+{
+    public class OsuLinkFlowContainer : OsuLinkFlowContainer<OsuSpriteLink>
+    {
+        public OsuLinkFlowContainer(Action<SpriteText> defaultCreationParameters = null)
+            : base(defaultCreationParameters)
+        {
+        }
+    }
+
+    public class OsuLinkFlowContainer<T> : OsuTextFlowContainer
+        where T : OsuSpriteLink, new()
+    {
+        public override bool HandleInput => true;
+
+        public OsuLinkFlowContainer(Action<SpriteText> defaultCreationParameters = null)
+            : base(defaultCreationParameters)
+        {
+        }
+
+        protected override SpriteText CreateSpriteText() => new T();
+
+        /// <summary>
+        /// The colour for text (links override this). Will only be used for new text elements.
+        /// </summary>
+        public ColourInfo TextColour = Color4.White;
+
+        public IEnumerable<SpriteText> AddLink(string text, string url, Action<SpriteText> creationParameters = null)
+        {
+            // TODO: Remove this and get word wrapping working
+            text = text.Replace(' ', '_');
+
+            return AddText(text, link =>

```

Comment:
rather than calling `AddText` and making every single `SpriteText` be an `OsuLink`, why not just do the link construction here? Means you won't need all this casting, default colour logic, generic T classes etc.

Reply:
This exists so that the link in `ProfileHeader` still works. Changing this class to creating the links itself (and to remove generics) would require a way of differentiating between general osu links and the more specific chat links. It would be easy to just put two different methods there or inherit this from another class just for chat links, but I'm really not sure what the preferred solution for this would be.

Reply:
I like the idea, but creating the sprite inside this means I can't use the `base.Add...` methods since all of them create the SpriteTexts based on the input text. If I want to do this differently I'd have to change `TextFlowContainer.Add` to not throw an exception anymore, and that doesn't seem intended. But if I changed it so we would _not_ let this class inherit from `TextFlowContainer`, a lot of the code would probably have to be rewritten